### PR TITLE
Merge 3.3 to main... Maybe?

### PR DESCRIPTION
**Pull request by hpidcock**
Might need to redo this one.

Forward ports:
- #16179
- #16122
- #16180
- #16182
- #16183
- #15969
- #16164
- #16192
- #16186
- #16200
- #16203
- #16204
- #16209
- #16033
- #16210
- #15878
- #15882
- #16196
- #16220

Conflicts:
- .github/workflows/build.yml
- .github/workflows/smoke.yml
- agent/agent.go
- agent/format-2.0.go
- apiserver/common/modelstatus_test.go
- apiserver/facades/agent/agent/agent_test.go
- apiserver/facades/agent/agent/model_test.go
- apiserver/facades/agent/agent/register.go
- apiserver/facades/agent/caasadmission/register.go
- apiserver/facades/agent/caasagent/register.go
- apiserver/facades/agent/provisioner/provisioner.go
- apiserver/facades/agent/secretsmanager/register.go
- apiserver/facades/client/controller/controller.go
- apiserver/facades/client/controller/controller_test.go
- apiserver/facades/controller/caasmodelconfigmanager/register.go
- apiserver/facades/controller/firewaller/firewaller_base_test.go
- apiserver/facades/controller/firewaller/register.go
- cmd/containeragent/initialize/config.go
- cmd/juju/application/bundle_test.go
- cmd/juju/application/deploy_test.go
- cmd/jujud/agent/agenttest/agent.go
- go.mod
- go.sum
- tests/suites/smoke/deploy.sh
- worker/uniter/runner/context/context_test.go
- worker/upgradedatabase/mocks/agent.go

### CODE REVIEW

File: .github/workflows/upgrade.yml [hunk 1]
```diff
@@ -46,9 +42,7 @@ jobs:
         shell: bash
         run: |
           set -euxo pipefail
-          sudo snap install snapcraft --classic
-          sudo snap install yq
-          sudo snap install juju --channel=3.1/stable
+          sudo snap install juju --channel=3.2/stable
```
Comment (barrettj12) [inline .github/workflows/upgrade.yml:45]: should this be 3.3?

File: state/testing/watcher.go [hunk 2]
```diff
@@ -92,6 +92,22 @@ loop:
 	c.AssertNoChange()
 }
 
+func (c NotifyWatcherC) AssertAtleastOneChange() {
```
Comment (barrettj12) [inline state/testing/watcher.go:95]: should be `AssertAtLeastOneChange`, capital L

**Review by barrettj12 — APPROVED** [review]
LGTM, probably worth getting others to review too

**Review by wallyworld — APPROVED** [review]

File: .github/workflows/upgrade.yml (hunk 1 above)
Comment (hpidcock) [reply .github/workflows/upgrade.yml:45]: 3.3 isn't released.

File: state/testing/watcher.go (hunk 2 above)
Comment (hpidcock) [reply state/testing/watcher.go:95]: I'll change it in 3.1.

**Pull request merged**